x-pack/plugins/cloud_security_posture/public/pages/vulnerabilities/hooks/use_latest_vulnerabilities_by_resource.ts

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { useQuery } from '@tanstack/react-query';
+import { lastValueFrom } from 'rxjs';
+import type { IKibanaSearchRequest, IKibanaSearchResponse } from '@kbn/data-plugin/common';
+import {
+  SearchRequest,
+  SearchResponse,
+  AggregationsCardinalityAggregate,
+  AggregationsMultiBucketAggregateBase,
+  AggregationsSingleBucketAggregateBase,
+  AggregationsStringRareTermsBucketKeys,
+  AggregationsStringTermsBucketKeys,
+  SortOrder,
+} from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import { LATEST_VULNERABILITIES_INDEX_PATTERN } from '../../../../common/constants';
+import { MAX_FINDINGS_TO_LOAD } from '../../../common/constants';
+import { useKibana } from '../../../common/hooks/use_kibana';
+import { showErrorToast } from '../../../common/utils/show_error_toast';
+import { FindingsBaseEsQuery } from '../../../common/types';
+import { getVulnerabilityFilter } from '../utils';
+
+type LatestFindingsRequest = IKibanaSearchRequest<SearchRequest>;
+type LatestFindingsResponse = IKibanaSearchResponse<SearchResponse<any, VulnerabilitiesAggs>>;
+
+interface VulnerabilitiesAggs {
+  count: AggregationsMultiBucketAggregateBase<AggregationsStringRareTermsBucketKeys>;
+  total: AggregationsCardinalityAggregate;
+  resources: AggregationsMultiBucketAggregateBase<FindingsAggBucket>;
+}
+
+interface FindingsAggBucket extends AggregationsStringRareTermsBucketKeys {
+  name: AggregationsMultiBucketAggregateBase<AggregationsStringTermsBucketKeys>;
+  region: AggregationsMultiBucketAggregateBase<AggregationsStringTermsBucketKeys>;
+  vulnerability_severity_critical: AggregationsSingleBucketAggregateBase;
+  vulnerability_severity_high: AggregationsSingleBucketAggregateBase;
+  vulnerability_severity_medium: AggregationsSingleBucketAggregateBase;
+  vulnerability_severity_low: AggregationsSingleBucketAggregateBase;
+}
+
+interface VulnerabilitiesQuery extends FindingsBaseEsQuery {
+  sortOrder: SortOrder;
+  enabled: boolean;
+  pageIndex: number;
+  pageSize: number;
+}
+
+export const getQuery = ({
+  query,
+  sortOrder = 'desc',
+  pageIndex,
+  pageSize,
+}: VulnerabilitiesQuery) => ({
+  index: LATEST_VULNERABILITIES_INDEX_PATTERN,
+  query: {
+    ...query,
+    bool: {
+      ...query?.bool,
+      filter: [...(query?.bool?.filter || []), ...getVulnerabilityFilter()],
+    },
+  },
+  aggs: {
+    total: { cardinality: { field: 'resource.id' } },
+    resources: {
+      terms: {
+        field: 'resource.id',
+        order: {
+          _count: 'desc',
+        },
+        size: MAX_FINDINGS_TO_LOAD,
```

Comment:
After digging into some research on the optimal size value, I reached the conclusion that just increasing it to the maximum will reduce performance drastically once the user passes 3000 unique resources. To address that issue, I reached into a better formula for the specific use case of vulnerability management:

I updated it to reflect the logic in the code, I increased the size to `1500`, (leaving it at `MAX_FINDINGS_TO_LOAD * 3`), that's the maximum without performance degrading, and then updated the sorting logic to be now:
 
Count -> Severity Critical -> Severity High -> Severity Medium.

This means, in descending order (default), it wil capture the top 1500 resources with more vulnerabilities and for tie-breaker criteria in case there are resources with same ammount of vulnerability it will consider the ones with the most severe vulnerabilities first to show in the top 500 table.

In the ascending order, it will get the top 1500 resources with less vulnerability while at the same time looking for the most severe ones first to collect to the 1500 resources bucket, meaning we are showing relevant data for the users on the descending sort as well.

The tiebreaking criteria:

<img width="1244" alt="image" src="https://github.com/elastic/kibana/assets/19270322/c43ae68a-1215-4987-8ce6-96179b0ad200">

Ascending order:

<img width="1255" alt="image" src="https://github.com/elastic/kibana/assets/19270322/62c301be-9f31-406f-98bf-7b480a71dda1">